File: app/src/main/java/io/digitallibrary/reader/SelectLanguageActivity.java
package io.digitallibrary.reader;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;

import io.digitallibrary.reader.utilities.LanguageUtil;

public class SelectLanguageActivity extends AppCompatActivity {
    public static final String TAG = "SelectLanguageActivity";
    private RelativeLayout background;
    private LinearLayout languageDrawer;
    private RelativeLayout spinnerAndErrorContainer;
    private ProgressBar spinner;
    private TextView errorMsg;
    private ArrayAdapter<String> langItemsAdapter;
    private ListView langList;
    private int selected;

    @Override
    protected void onCreate(final Bundle state) {
        super.onCreate(state);
        setContentView(R.layout.select_language);
        background = findViewById(R.id.background);
        languageDrawer = findViewById(R.id.language_drawer);
        spinnerAndErrorContainer = findViewById(R.id.spinner_and_error_msg);
        spinner = findViewById(R.id.spinner);
        errorMsg = findViewById(R.id.error_msg);
        langList = findViewById(R.id.languages);
        overridePendingTransition(0,0);
        ViewTreeObserver vto = languageDrawer.getViewTreeObserver();
        background.setAlpha(0f);
        vto.addOnGlobalLayoutListener (new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                animateIn();
            }
        });

        fetchLanguages();
    }

    private void fetchLanguages() {
        LanguageUtil.getLanguageListFromServer(this, new LanguageUtil.Callback() {
            @Override
            public void onSuccess(final Map<String, String> languages) {
                final ArrayList<String> langArray = new ArrayList<>(languages.keySet());
                Collections.sort(langArray);
                selected = -1;
                for (int i=0; i != langArray.size(); ++i) {
                    Log.i(TAG, "i " + i);
                    if (langArray.get(i).equals(LanguageUtil.getCurrentLanguageText())) {
                        selected = i;
                        continue;
                    }
                }
                Log.i(TAG, "Selected is: " + selected);
                langItemsAdapter = new ArrayAdapter<String>(SelectLanguageActivity.this, R.layout.language_row, R.id.language_name, langArray) {
                    @NonNull
                    @Override
                    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                        View v = super.getView(position, convertView, parent);
                        if (position == selected) {
                            v.setSelected(true);
                            v.findViewById(R.id.checked).setVisibility(View.VISIBLE);
                            v.setBackgroundResource(R.drawable.nav_menu_background_pressed);
                        } else {
                            v.findViewById(R.id.checked).setVisibility(View.INVISIBLE);
                            v.setBackgroundResource(R.drawable.nav_menu_background_default);
                        }
                        return v;
                    }
                };
                spinnerAndErrorContainer.setVisibility(View.GONE);
                langList.setVisibility(View.VISIBLE);
                langList.setAdapter(langItemsAdapter);
                langList.setItemChecked(selected, true);
                langList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        Log.i(TAG, "onItemClick: " + position);
                        selected = position;
                        String newLang = langArray.get(position);
                        String newLangCode = languages.get(newLang);
                        String oldLang = LanguageUtil.getCurrentLanguage();
                        LanguageUtil.setLanguage(newLangCode, newLang);
                        view.findViewById(R.id.checked).setVisibility(View.VISIBLE);
                        langItemsAdapter.notifyDataSetChanged();
                        if (!oldLang.equals(newLang)) {
                            Gdl.fetch();
                        }
                    }
                });
            }

            @Override
            public void onError() {
                spinner.setVisibility(View.GONE);
                errorMsg.setVisibility(View.VISIBLE);
            }
        });
    }

    private void animateIn() {
        final float scale = getApplication().getResources().getDisplayMetrics().density;
        languageDrawer.animate().x(47f * scale);
        background.animate().alpha(0.3f);
    }

    private void animateOut() {
        final float scale = getApplication().getResources().getDisplayMetrics().density;
        languageDrawer.animate().x(-267f * scale);
        background.animate().alpha(0f);
    }

    @Override
    public void onBackPressed() {
        animateOut();
        languageDrawer.animate().withEndAction(new Runnable() {
            @Override
            public void run() {
                SelectLanguageActivity.super.onBackPressed();
                SelectLanguageActivity.this.overridePendingTransition(0,0);
            }
        });
    }

    public void onCloseLanguagesClicked(View view) {
        animateOut();
        languageDrawer.animate().withEndAction(new Runnable() {
            @Override
            public void run() {
                SelectLanguageActivity.this.finish();
                SelectLanguageActivity.this.overridePendingTransition(0,0);
            }
        });
    }

    public void onBackgroundClicked(View view) {
        animateOut();
        languageDrawer.animate().withEndAction(new Runnable() {
            @Override
            public void run() {
                SelectLanguageActivity.this.finish();
                SelectLanguageActivity.this.overridePendingTransition(0,0);
            }
        });
    }
}

File: app/src/main/java/io/digitallibrary/reader/SelectLanguageActivity.kt
package io.digitallibrary.reader

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.TextView
import io.digitallibrary.reader.catalog.Language
import io.digitallibrary.reader.utilities.SelectionsUtil
import kotlinx.android.synthetic.main.activity_select_language.*
import kotlinx.android.synthetic.main.toolbar.*
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch

class SelectLanguageActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "SelectLanguageActivity"
        const val LANGUAGE_SELECTED = "LANGUAGE_SELECTED"
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreate(state: Bundle?) {
        setTheme(Gdl.getSettingsThemeId())
        super.onCreate(state)
        setContentView(R.layout.activity_select_language)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setTitle(R.string.languages_selection_title)

        // -1 means not set
        // Correct value will be set when languages is added to the adapter
        var selectedLangPosition = -1

        val langItemsAdapter = object : ArrayAdapter<Language>(this@SelectLanguageActivity, R.layout.item_language_row, R.id.language_name) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val v = super.getView(position, convertView, parent)
                if (position == selectedLangPosition) {
                    v.isSelected = true
                    v.findViewById<View>(R.id.checked).visibility = View.VISIBLE
                    v.findViewById<TextView>(R.id.language_name).setTextColor(ContextCompat.getColor(context, R.color.gdl_green))
                    v.findViewById<TextView>(R.id.language_name).text = getItem(position).languageName
                } else {
                    v.findViewById<View>(R.id.checked).visibility = View.INVISIBLE
                    v.findViewById<TextView>(R.id.language_name).setTextColor(ContextCompat.getColor(context, R.color.dark_text))
                    v.findViewById<TextView>(R.id.language_name).text = getItem(position).languageName
                }
                return v
            }
        }

        language_list.adapter = langItemsAdapter

        val shortDuration = resources.getInteger(android.R.integer.config_shortAnimTime).toLong()
        ViewModelProviders.of(this).get(LanguagesViewModel::class.java).languages.observe(this, Observer {
            it?.let {
                val fadeFromView = if (it.isEmpty()) language_list else spinner
                val fadeToView = if (it.isEmpty()) spinner else language_list

                fadeFromView.animate().alpha(0f).setDuration(shortDuration).setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator?) {
                        super.onAnimationEnd(animation)
                        fadeFromView.visibility = View.GONE
                    }
                })

                fadeToView.visibility = View.VISIBLE
                fadeToView.alpha = 0f
                fadeToView.animate().alpha(1f).setDuration(shortDuration).setListener(null)

                val currentLink = SelectionsUtil.getCurrentLanguageLink()
                selectedLangPosition = it.indexOfFirst { it.link == currentLink }

                langItemsAdapter.clear()
                langItemsAdapter.addAll(it)
                langItemsAdapter.notifyDataSetChanged()

                // Need to wait for the view to be drawn, before we know if we have to scroll
                val layoutListener = object : ViewTreeObserver.OnGlobalLayoutListener {
                    override fun onGlobalLayout() {
                        language_list.viewTreeObserver.removeOnGlobalLayoutListener(this)
                        if (selectedLangPosition != -1) {
                            if (language_list.lastVisiblePosition < selectedLangPosition + 2) {
                                language_list.smoothScrollToPosition(selectedLangPosition + 2)
                            } else if (language_list.firstVisiblePosition > selectedLangPosition - 2) {
                                language_list.smoothScrollToPosition(selectedLangPosition - 2)
                            }
                        }
                    }
                }
                language_list.viewTreeObserver.addOnGlobalLayoutListener(layoutListener)
            }
        })

        language_list.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            SelectionsUtil.setLanguage(langItemsAdapter.getItem(position).link, langItemsAdapter.getItem(position).languageName)
            // Try to get default category, if we don't have any yet, the OPDS parser will set one
            launch(UI) {
                val categories = async { Gdl.database.categoryDao().getCategories(langItemsAdapter.getItem(position).link) }.await()
                if (categories.isNotEmpty()) {
                    SelectionsUtil.setCategory(categories[0].link, categories[0].title)
                }
            }
            Gdl.fetchOpdsFeed()
            val intent = Intent(LANGUAGE_SELECTED)
            LocalBroadcastManager.getInstance(Gdl.appContext).sendBroadcast(intent)
            finish()
        }
    }
}